com.unity.shadergraph/Editor/Data/Nodes/NodeClassCache.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Profiling;
+
+namespace UnityEditor.ShaderGraph
+{
+    [InitializeOnLoad]
+    public static class NodeClassCache
+    {
+        private static Dictionary<Type,List<ContextFilterableAttribute>> m_KnownTypeLookupTable;
+
+        private static Dictionary<Type, List<ContextFilterableAttribute>> KnownTypeLookupTable
+        {
+            get
+            {
+                if(m_KnownTypeLookupTable == null)
+                {
+                    ReCacheKnownNodeTypes();
+                }
+                return m_KnownTypeLookupTable;
+            }
+        }
+
+        public static IEnumerable<Type> KnownNodeTypes
+        {
+            get => KnownTypeLookupTable.Keys;
+        }
+
+        public static IEnumerable<ContextFilterableAttribute> GetFilterableAttributesOnNodeType(Type nodeType)
+        {
+            if (nodeType == null)
+            {
+                throw new ArgumentNullException("Cannot get attributes on a null Type");
+            }
+
+            if (KnownTypeLookupTable.TryGetValue(nodeType, out List<ContextFilterableAttribute> filterableAttributes))
+            {
+                return filterableAttributes;
+            }
+            else
+            {
+                throw new ArgumentException($"The passed in Type {nodeType.FullName} was not found in the loaded assemblies as a child class of AbstractMaterialNode");
+            }
+        }
+
+        public static T GetAttributeOnNodeType<T>(Type nodeType) where T : ContextFilterableAttribute
+        {
+            var filterableAttributes = GetFilterableAttributesOnNodeType(nodeType);
+            foreach(var attr in filterableAttributes)
+            {
+                if(attr is T searchTypeAttr)
+                {
+                    return searchTypeAttr;
+                }
+            }
+            return null;
+        }
+
+        private static void ReCacheKnownNodeTypes()
+        {
+            Profiler.BeginSample("NodeClassCache: Re-caching all known node types");
+            m_KnownTypeLookupTable = new Dictionary<Type, List<ContextFilterableAttribute>>();
+            foreach (var assem in AppDomain.CurrentDomain.GetAssemblies())
+            {
+                foreach (Type nodeType in assem.GetTypes())
```

Comment:
This is going to have an issue with DOTS..  some assemblies throw exceptions when you call GetTypes().
See fix here:   https://github.com/Unity-Technologies/Graphics/pull/910/commits/228d2f8041bce45d2b636b354bcc3cbc3db39374

Reply:
Also should probably just use typecache instead:   https://forum.unity.com/threads/unityeditor-typecache-api-for-fast-extraction-of-type-attributes-in-the-editor-tooling.687682/

Reply:
Changed to typecache, should now only call GetType on Attributes now